core/src/martian/encoders.cljc

```diff
@@ -1,40 +1,94 @@
 (ns martian.encoders
-  (:require [clojure.string :as string]
-            [clojure.walk :refer [keywordize-keys]]
+  (:require [clojure.string :as str]
             [cognitect.transit :as transit]
             [flatland.ordered.map :refer [ordered-map]]
             #?(:clj [cheshire.core :as json])
-            #?(:clj [clojure.edn :as edn]
+            #?(:clj  [clojure.edn :as edn]
                :cljs [cljs.reader :as edn])
             #?(:clj [clojure.java.io :as io])
             #?(:clj [martian.multipart :as multipart])
-            #?@(:bb []
+            #?@(:bb  []
                 :clj [[ring.util.codec :as codec]]))
-  #?(:clj (:import [java.io ByteArrayInputStream ByteArrayOutputStream])))
+  #?(:clj (:import [java.io ByteArrayInputStream ByteArrayOutputStream InputStream PushbackReader])))
 
-(defn transit-decode [body type]
-  #?(:clj (transit/read (transit/reader (ByteArrayInputStream. body) type))
-     :cljs (transit/read (transit/reader type {}) body)))
+;; NB: THIS NS ASSUMES THAT "UTF-8" IS THE CHARSET OF ALL ENCODED/DECODED VALUES!
 
-(defn transit-encode [body type]
-  #?(:clj
-     (let [out (ByteArrayOutputStream. 4096)
-           writer (transit/writer out type)]
-       (transit/write writer body)
-       (io/input-stream (.toByteArray out)))
+#?(:clj
+   (defn as-bytes [obj]
+     (cond
+       (bytes? obj) obj
+
+       (string? obj)
+       (.getBytes ^String obj)
+
+       (instance? InputStream obj)
+       (with-open [^InputStream is obj
+                   baos (ByteArrayOutputStream.)]
+         (io/copy is baos)
+         (.toByteArray baos))
+
+       ;; Leave as is and let it fail downstream
+       :else obj)))
+
+#?(:clj
+   (defn as-stream [obj]
+     (if (instance? InputStream obj)
+       obj
+       (ByteArrayInputStream. (as-bytes obj)))))
+
+#?(:clj
+   (defn as-string
+     ([obj]
+      (if (string? obj)
+        obj
+        (as-string obj "UTF-8")))
+     ([obj ^String charset]
+      (if (string? obj)
+        obj
+        (slurp obj :encoding charset)))))
```

Comment:
The decoder fns are now more flexible thanks to `as-<raw-type>` coercion fns.
Their design was influenced by existing `clj-http`/`hato` utils implementation.